datafusion/physical-expr/src/intervals/interval_aritmetic.rs

```diff
@@ -508,6 +580,83 @@ fn cast_scalar_value(
     ScalarValue::try_from_array(&cast_array, 0)
 }
 
+/// This function calculates the final cardinality result by inspecting the endpoints of the interval.
+fn calculate_cardinality_based_on_bounds(
+    lower_open: bool,
+    upper_open: bool,
+    diff: u64,
+) -> u64 {
+    match (lower_open, upper_open) {
+        (false, false) => diff + 1,
+        (true, true) => diff - 1,
+        _ => diff,
+    }
+}
+
+trait OneTrait: Sized + std::ops::Add + std::ops::Sub {
+    fn one() -> Self;
+}
+
+macro_rules! impl_OneTrait{
+    ($($m:ty),*) => {$( impl OneTrait for $m  { fn one() -> Self { 1 as $m } })*}
+}
+impl_OneTrait! {u8, u16, u32, u64, i8, i16, i32, i64, f32, f64}
+
+/// This function either increments or decrements its argument, depending on the `DIR` value. If `true`, it increments; otherwise it decrements the argument.
+fn increment_decrement<const DIR: bool, T: OneTrait + SubAssign + AddAssign>(
+    mut val: T,
+) -> T {
+    if DIR {
+        val.add_assign(T::one());
+    } else {
+        val.sub_assign(T::one());
+    }
+    val
+}
+
+/// This function returns the next/previous value depending on the `DIR` value.
+/// If `true`, it returns the next value; otherwise it returns the previous value.
+fn get_next_value<const DIR: bool>(value: ScalarValue) -> ScalarValue {
+    use ScalarValue::*;
+    match value {
+        Float32(Some(val)) => {
+            let incremented_bits = increment_decrement::<DIR, u32>(val.to_bits());
+            Float32(Some(f32::from_bits(incremented_bits)))
+        }
+        Float64(Some(val)) => {
+            let incremented_bits = increment_decrement::<DIR, u64>(val.to_bits());
+            Float64(Some(f64::from_bits(incremented_bits)))
+        }
+        Int8(Some(val)) => Int8(Some(increment_decrement::<DIR, i8>(val))),
+        Int16(Some(val)) => Int16(Some(increment_decrement::<DIR, i16>(val))),
+        Int32(Some(val)) => Int32(Some(increment_decrement::<DIR, i32>(val))),
+        Int64(Some(val)) => Int64(Some(increment_decrement::<DIR, i64>(val))),
+        UInt8(Some(val)) => UInt8(Some(increment_decrement::<DIR, u8>(val))),
+        UInt16(Some(val)) => UInt16(Some(increment_decrement::<DIR, u16>(val))),
+        UInt32(Some(val)) => UInt32(Some(increment_decrement::<DIR, u32>(val))),
+        UInt64(Some(val)) => UInt64(Some(increment_decrement::<DIR, u64>(val))),
+        _ => value, // Infinite bounds or unsupported datatypes
+    }
+}
+
+/// This function takes an interval, and if it has any open bound(s), it
+/// converts them to closed bound(s) preserving the interval endpoints.
+pub fn interval_with_closed_bounds(mut interval: Interval) -> Interval {
```

Comment:
This might be more elegant if it was a method on  a `Interval` itelf

```rust
impl Interval 
  pub fn iwith_closed_bounds(mut self -> Interval {
...
}
}
```